openid-4-verifiable-credential-issuance-1_0.md

```diff
@@ -1119,6 +1121,96 @@ Cache-Control: no-store
 }
 ```
 
+# Callback Endpoint {#callback_endpoint}
+
+This endpoint is used by the Wallet to notify the Credential Issuer whether a Credential has been successfully received or not. The Credential Issuer needs to return `callback_id` in the Credential Response or a Batch Credential Response for the Wallet to be able to use this Endpoint. Support for this endpoint is OPTIONAL. The wallet MUST call this endpoint if the Credential Issuer supports it and provides a `callback_id`.
+
+This endpoint can be used after the Credential Issuer has sent Credential Response or Batch Credential Response. It enables the Credential Issuer to take subsequent actions after issuance, depending on whether the Credential has been accepted and successully stored by the Wallet, rejected by the Wallet, or errors and other unforeseen circumstances have occurred during the Wallet's processing.
+
+The Wallet MUST present to the Callback Endpoint a valid Access Token issued at the Token Endpoint as defined in (#token_endpoint). Credential Issuer that requires request to the Callback Endpoint MUST ensure Access Token issued by the Authorization Server is valid at the Callback Endpoint.
+
+The callback from the Wallet is idempotent. The Credential Issuer MUST return success if it receives multiple identical calls from the Wallet for the same `callback_id`s.
+
+The Wallet MAY retry if the call to this endpoint fails for a temporary reason. The Credential Issuer SHOULD pre-determine the amount of time within which it expects the callback. Therefore, even a well-formed callback from the Wallet could fail if received by the Credential Issuer after this time period. The Credential Issuer may never receive the callback, meaning it is unknown whether the Wallet successfully stored the credential or not - it is left to the Credential Issuer to decide how to proceed in this case.
```

Comment:
```suggestion
The Wallet MUST retry if the call to this endpoint fails for a temporary reason. The Credential Issuer SHOULD pre-determine the amount of time within which it expects the callback. Therefore, even a well-formed callback from the Wallet could fail if received by the Credential Issuer after this time period. The Credential Issuer may never receive the callback, meaning it is unknown whether the Wallet successfully stored the credential or not - it is left to the Credential Issuer to decide how to proceed in this case.
```
The implications of MAY are as follows. 
If the wallet does not have to retry in the case of failure, then the issuer is in no better a position than if this endpoint was not used in the first place. And this text provides no guidance on what the issuer should do next. Furthermore it does not indicate how the user experience will be different whether this endpoint is called or not.

If the issuer assumes success, then it could assume success without the need for this endpoint.
If the issuer assumes failure, is it going to revoke the user's credential which is safely stored in the wallet? What sort of user experience would this give?

For these reasons, without a properly documented retry and recovery procedure, I do not believe that this endpoint adds any significant benefit because the decisions that need to be made in the event of a failure at this endpoint are similar to the ones the wallet and issuer must make if the initial credential issuance failed for some reason.

Reply:
> If the wallet does not have to retry in the case of failure, then the issuer is in no better a position than if this endpoint was not used in the first place.

I don't think that's true. The experience for 95%+ of users is still better.

> For these reasons, without a properly documented retry and recovery procedure, I do not believe that this endpoint adds any significant benefit

There's a fair number of implementors that believe otherwise. I think if you have some suggestions on improving they'd be welcome, but I think any such changes may be mostly editorial as I don't think we should add any 'MUST' for issuer behaviour in this area.

I'm not sure whether the 'MAY' should be a 'MAY', 'SHOULD' or 'MUST'. That is probably worthy of a wider discussion, but probably isn't a reason to hold up this PR.

Reply:
If you have a good connection then the initial credential delivery is highly unlikely to fail. So you do not need the proposed callback endpoint.
If you have a very flaky connection then the initial credential delivery is highly likely to fail, but so is the callback confirmation. So again you have not provided any significant benefit.
I suggest it's only in the intermediate cases where say 1 in 100 fail that this would increase to 1 in 10,000 by using the callback. But again, I do not see the benefit in implementing it. 
In our implementation we never had it and the user experience was not detrimental from not having it.
Also I am not sure why you are saying that the provision of a callback endpoint actually affects the GUI, since the  user should be protected from the communications exchanges.

Reply:
It's not just about the network connection, there's a whole bunch of things that can go wrong.

> Also I am not sure why you are saying that the provision of a callback endpoint actually affects the GUI, since the user should be protected from the communications exchanges.

I don't think I follow this, can you expand please?

One example use case is the cross device case, where the user is at the issuer's website on their desktop installing a credential into a mobile wallet. It's generally helpful if the issuer website can update once the user has successfully installed the credential, for example giving the user guidance as to the things they might want to do next.

Reply:
Can you please explain what you mean by "a whole bunch of things can go wrong". What are these things? I think it would help to determine the rationale for this new endpoint if we knew what problems we are trying to solve.

Concerning the GUI, the user requests the credential, and the wallet communicates with the issuer and installs the credential.  The communications do not need any user interaction, except when the user must authenticate to prove possession of the private key and sign the JWT with the public key. But this is before the new endpoint is called, so it is out of scope of the current discussion. After the credential is received, the user sees the credential in the wallet, or does not if it has failed. So the user knows the final state because he holds the wallet. If the issuer wants the user to further interact with it after the credential installation (which I do not think will be the majority of cases - certainly we never had any need for this - then its web page (that could be the one displaying the QR code that kicks off the installation) can display "click here once you have received the credential". This provides the equivalent of the new endpoint for cases where it is needed.

Reply:
> Can you please explain what you mean by "a whole bunch of things can go wrong". What are these things

Off the top of my head: The wallet can crash, the phone can run out of battery, the user can correctly scan the qr code and think they've finished the process, the user can take a photo of the QR code and think they're done, the user can press the 'reject' button accidentally, the user can get distracted, the credential can fail to validate for some reason and get rejected by the wallet or the user somehow ends up in a wallet that doesn't support the credential, writing the credential to the phone storage can fail due to space restrictions, or due to a faulty sd card, the credential can turn out to be a different one that the user was expecting, the credential can fail to meet the wallet policy for other reasons, there can be a failure at the wallet backend.

I think we are at least in agreement that having the endpoint enables more choice in the UX?

If so the debate may be over whether a UX that puts more cognitive load onto the user is a better or worse UX, or at least whether that is worth the tradeoff of additional complexity in the protocol/wallet?

Reply:
I don't think any of the errors you have suggested are a good enough reason for this new endpoint. You simply need the wallet to show the user the new credential once it has been successfully received and the issuer can assume all is OK. If it fails for any reason the user will be aware and can start the transaction again. The issuer will be aware of the failure because the user is retrying to install the credential. If the user cannot be bothered to retry, the issuer will think all is OK when it is not. But this is no different to the user getting the credential and never using it. Neither is different from the issuer's perspective or the user's perspective.

So there is no significant benefit from this new endpoint to either party and it unnecessarily complicates the protocol.

The new endpoint has no effect on the user experience and GUI.

Reply:
> the issuer can assume all is OK

No, they can't, and I think this is fundamental to this discussion. All user experience design has to essentially start from a basis that many users will not properly understand this flow the first time they do it and will get confused by it. Any improvement that reduces the chances of the user having an incorrect understanding has massive benefits for any kind of at scale deployment. The flows you are suggesting all rely on the user doing more things correctly (and reduce the ability of the issuer to guide the user) compared to the flows I am suggesting.

> The new endpoint has no effect on the user experience and GUI.

I'm not sure if you and I are using the 'user experience' term to mean difference things. I'll try without using that term. You say the GUI is affected in https://github.com/openid/OpenID4VCI/pull/70#discussion_r1396222809 where you describe how you can implement an alternative (and arguably worse) GUI instead of using this endpoint. If I can't implement the GUI I describe without this endpoint, then the presence/absence of this endpoint affects the user experience the issuer can provide.

If you think your suggested user experience is the preferable one, that's fine, that's your opinion and you are welcome to implement your issuer like that. But the protocol should not unnecessarily prevent other people implementing user experiences they think (or have proved through user research) are better.

Reply:
Can you please explain how your user experience or GUI is improved by this new endpoint. Thanks.

Reply:
Since all that it is doing (from my perspective) is telling the issuer what has happened. The GUI at the wallet end will be the same AFAIK whether this new endpoint is contacted or not.

Reply:
I explained that in https://github.com/openid/OpenID4VCI/pull/70#discussion_r1396166406

Reply:
You said  "Any improvement that reduces the chances of the user having an incorrect understanding has massive benefits for any kind of at scale deployment." Please explain why telling the issuer the outcome of the transaction helps the user gain a more correct understanding, since the wallet already knows what the outcome is!!!

Reply:
This comment 
https://github.com/openid/OpenID4VCI/pull/70#discussion_r1396166406
is not sufficient in my opinion to add a new exchange to the protocol (one that is flawed anyway as I explained in my review requesting MAY should be MUST) since the web site can simply ask the user to click on a link once the credential has been correctly installed.

Reply:
Furthermore you are mandating that wallets implement it if the issuer requests it.

Reply:
So you are placing a burden on all wallet implementors for those few web sites that may wish to design their web site in one specific way.

Reply:
As I mentioned in https://github.com/openid/OpenID4VCI/pull/70#discussion_r1396166406 in that flow the user is still interacting with the issuer at the end of the cross device flow.

The wallet often does not have the context of why a credential was issued to the user or what can be done with the credential; the issuer may well.

We're going round in circles now; we already addressed your latest point in https://github.com/openid/OpenID4VCI/pull/70#discussion_r1397138417 - it is frankly completely daft to suggest we should limit ourselves to a UX that relies on the user to understand whether or not they correctly installed the credential. If that's the UX you want to implement in your systems, that's fine - but a number of people that have much better understanding of UX and user behaviour than I do believe otherwise, including the person that originally opened https://github.com/openid/OpenID4VCI/issues/32 - and I tend to agree with them all.

Reply:
You are mandating that all wallet implementors must implement this feature because they do not know if an issuer will request it or not. As a previous wallet implementor I strongly object to this added complexity.

Reply:
David, one of the incentives for this endpoint is not only to improve the UX, but also to add a guarantee to the issuer that the credential has been issued. Issuers using third party providers to issue credentials really need it because they do not want to be charged for the credentials, whose issuance was not successful for various reasons that Joseph and David Luna has been outlining.

multiple companies have a variant of this endpoint implemented and having it standardized really helps the ecosystem.

Reply:
Mandating the wallet to support this optional endpoint when the issuer supports it could be seen as extra work for the wallet. But the wallet can choose not to interact with the issuer who supports this endpoint (based on the issuer metadata), or the wallet can simplify the implementation by for example not retrying the callback if it fails once. so the wallet has a lot of discretion.

This is also why, while making it optional to the wallet to retry the callback when the callback itself is mandatory when the issuer supports it could look insufficient at first glance, it is also a way to simplify the wallet implementations.

Reply:
> Issuers using third party providers to issue credentials really need it because they do not want to be charged for the credentials, whose issuance was not successful for various reasons that Joseph and David Luna has been outlining.

I would like to better understand this kind of setup. I feel the proposed mechanisms might not robust enough to build a charging solution on top of it.

Reply:
I do not believe that third party providers are part of the 3 party model are they? I have not seen any architectural diagrams that include them, so why are we proposing to alter the issuer holder protocol to cater for them?